### PR TITLE
Improve pibt

### PR DESCRIPTION
**Pull request by NoyGabay**

### CODE REVIEW

File: src/main/java/BasicMAPF/Solvers/PIBT/PIBT_Solver.java [hunk 1]
```diff
@@ -92,63 +111,116 @@ protected void init(MAPF_Instance instance, RunParameters parameters) {
...
+                            if (this.reachingGoalsConfigurations.get(i).get(agent).equals(this.reachingGoalsConfigurations.get(this.timeStamp).get(agent))) {
+                                // agent could be in a loop
+                                agentsInSameLocation++;
+                            }
+                        }
+                    }
+//                    System.out.println("Agents in same location: " + agentsInSameLocation);
+                    if (agentsInSameLocation == this.agentPlans.keySet().size()) {
+                        System.out.println("Loop detected in timestamp: " + i);
+                        System.out.println("current timestamp: " + this.timeStamp);
```
Comment (J-morag) [inline src/main/java/BasicMAPF/Solvers/PIBT/PIBT_Solver.java:153]:
```suggestion
                        if (DEBUG >= 2){
                            System.out.println("Loop detected in timestamp: " + i);
                            System.out.println("current timestamp: " + this.timeStamp);
                        }
```
Don't print by default. Add `private static final int DEBUG = 0;` at the beginning of the class. But you can add it to the instance report if you want.

File: src/main/java/BasicMAPF/Solvers/PIBT/PIBT_Solver.java [hunk 2]
```diff
@@ -92,63 +111,116 @@ protected void init(MAPF_Instance instance, RunParameters parameters) {
...
+//                        }
+
+                        break;
+                    }
+                }
+            }
+            if (loopDetected) {
+                System.out.println("LOOP DETECTED!!");
+                return null;
+            }
```
Comment (J-morag) [inline src/main/java/BasicMAPF/Solvers/PIBT/PIBT_Solver.java:198]: So this loop is very slow. Instead, hold a set of location configurations (`Set<List<I_Location>>`) and clear it whenever an agent becomes "finished". If a new configuration already exists in the set, we have a loop. 
Then I think you will no longer need `reachingGoalsConfigurations`.

File: src/main/java/BasicMAPF/Solvers/PIBT/PIBT_Solver.java [hunk 3]
```diff
@@ -182,11 +254,15 @@ protected boolean solvePIBT(Agent current, @Nullable Agent higherPriorityAgent)
         //  2. move to the node where the higher priority agent is
 
         // add current location of current agent - for the option to stay in current node in the next timestamp
-        if (higherPriorityAgent == null) {
-            candidates.add(this.currentLocations.get(current)); // 1
-        }
+//        if (higherPriorityAgent == null) {
+//            candidates.add(this.currentLocations.get(current)); // 1
+//        }
```
Comment (J-morag) [inline src/main/java/BasicMAPF/Solvers/PIBT/PIBT_Solver.java:259]:
```suggestion
```

File: src/main/java/BasicMAPF/Solvers/PIBT/PIBT_Solver.java [hunk 4]
```diff
@@ -182,11 +254,15 @@ protected boolean solvePIBT(Agent current, @Nullable Agent higherPriorityAgent)
...
         // prevent move to the node where the higher priority agent is
-        else {
+//        else {
+//            candidates.remove(this.currentLocations.get(higherPriorityAgent)); // 2
+//        }
```
Comment (J-morag) [inline src/main/java/BasicMAPF/Solvers/PIBT/PIBT_Solver.java:263]:
```suggestion
```

File: src/main/java/BasicMAPF/Solvers/PIBT/PIBT_Solver.java [hunk 5]
```diff
@@ -369,6 +446,32 @@ private boolean addNewMoveToAgent(Agent current, I_Location newLocation) {
...
+    private void updateStatus() {
+        this.reachingGoalsConfigurations.put(this.timeStamp, new HashMap<>());
+        for (Agent agent : this.priorities.keySet()) {
+            Boolean reachedGoal = (this.priorities.get(agent) == -1.0);
+            this.reachingGoalsConfigurations.get(this.timeStamp).put(agent, reachedGoal);
+        }
+    }
+
+    /**
+     * relevant function for lifelong version of MAPF.
```
Comment (J-morag) [inline src/main/java/BasicMAPF/Solvers/PIBT/PIBT_Solver.java:463]:
```suggestion
```

**Review by J-morag — CHANGES_REQUESTED** [review]
Good :)
Small comments mostly, shouldn't take too long to fix. Next time, please make sure to delete commented-out code beforehand.
Maybe add a small test where a loop should happen, and see that you detect it.

File: src/main/java/BasicMAPF/Solvers/PIBT/PIBT_Solver.java [hunk 6]
```diff
@@ -92,63 +113,61 @@ protected void init(MAPF_Instance instance, RunParameters parameters) {
...
             // init agents that have not reached their goal
             this.unhandledAgents = new HashSet<>();
             for (Map.Entry<Agent, Double> entry : this.priorities.entrySet()) {
                 Agent agent = entry.getKey();
-                Double priority = entry.getValue();
-                if (priority != -1.0) {
-                    // the agent did not reach his goal
-                    this.unhandledAgents.add(agent);
-                }
+                this.unhandledAgents.add(agent);
```
Comment (J-morag) [inline src/main/java/BasicMAPF/Solvers/PIBT/PIBT_Solver.java:154]: Are you sure this is OK?

File: src/main/java/BasicMAPF/Solvers/PIBT/PIBT_Solver.java [hunk 7]
```diff
@@ -92,63 +113,58 @@ protected void init(MAPF_Instance instance, RunParameters parameters) {
...
-            if (checkTimeout()) {
+            // loop detection
+            ArrayList<I_Location> currentConfiguration = new ArrayList<>(instance.agents.size());
+            for (Agent agent : instance.agents) {
+                currentConfiguration.add(this.currentLocations.get(agent));
+            }
+            if (this.configurations.contains(currentConfiguration)) {
+                if (DEBUG >= 2){
+                    System.out.println("LOOP DETECTED");
+                }
```
Comment (J-morag) [inline src/main/java/BasicMAPF/Solvers/PIBT/PIBT_Solver.java:138]: I pushed some changes here. Most importantly, note that you were iterating over a set to create the configuration, meaning the agents were not guaranteed to be the same order every time, so we could have had both false negatives and false positives.

**Review by J-morag — APPROVED** [review]
Approved, but see the two comments I left. And then you can merge when you're done with them.

File: src/main/java/BasicMAPF/Solvers/PIBT/PIBT_Solver.java (hunk 6 above)
Comment (NoyGabay) [reply src/main/java/BasicMAPF/Solvers/PIBT/PIBT_Solver.java:154]: its weird I changed it long time ago and i don't see those lines of code (not at my project and not in the git)

Comment (J-morag) [reply src/main/java/BasicMAPF/Solvers/PIBT/PIBT_Solver.java:154]: Must be from the mess I made with reverting stuff 😅

**Pull request merged**